crates/rattler/src/libsolv/pool.rs

```diff
@@ -0,0 +1,307 @@
+    pub fn create_whatprovides(&mut self) {
+        // Safe because pointer must exist
+        unsafe {
+            ffi::pool_createwhatprovides(self.as_ptr().as_mut());
+        }
+    }
+}
+
+/// Interns string like types into a `Pool` returning a `StringId`
+fn intern_str<T: AsRef<str>>(pool: &mut PoolRef, str: T) -> StringId {
+    // Safe because conversion is valid
+    let c_str = CString::new(str.as_ref()).expect("could never be null because of trait-bound");
+    let length = c_str.as_bytes().len();
+    let c_str = c_str.as_c_str();
+
+    // Safe because pool exists and function accepts any string
+    unsafe {
+        StringId(ffi::pool_strn2id(
+            pool.as_ptr().as_mut(),
+            c_str.as_ptr(),
+            length.try_into().expect("string too large"),
+            1,
+        ))
+    }
+}
+
+/// Finds a previously interned string or returns `None` if it wasn't found.
+fn find_intern_str<T: AsRef<str>>(pool: &PoolRef, str: T) -> Option<StringId> {
+    // Safe because conversion is valid
+    let c_str = CString::new(str.as_ref()).expect("could never be null because of trait-bound");
+    let length = c_str.as_bytes().len();
+    let c_str = c_str.as_c_str();
+
+    // Safe because pool exists and function accepts any string
+    unsafe {
+        let id = ffi::pool_strn2id(
+            pool.as_ptr().as_ptr(),
+            c_str.as_ptr(),
+            length.try_into().expect("string too large"),
+            0,
+        );
+        (id != 0).then(|| StringId(id))
+    }
+}
+
+/// Interns an instance of `Self` into a [`Pool`] returning a handle (or `Id`) to the actual data.
+/// Interning reduces memory usage by pooling data together which is considered to be equal, sharing
+/// the same `Id`. However, a `Pool` also only releases memory when explicitly asked to do so or on
+/// destruction.
+pub trait Intern {
+    type Id;
+
+    /// Interns the type in the [`Pool`]
+    fn intern(&self, pool: &mut PoolRef) -> Self::Id;
+}
+
+/// Enables retrieving the `Id` of previously interned instances of `Self` through the `Intern`
+/// trait.
+pub trait FindInterned: Intern {
+    /// Finds a previously interned instance in the specified [`Pool`]
+    fn find_interned_id(&self, pool: &PoolRef) -> Option<Self::Id>;
+}
+
+/// Wrapper for the StringId of libsolv
+#[derive(Copy, Clone)]
+pub struct StringId(pub(super) ffi::Id);
+
+impl StringId {
+    /// Resolve to the interned type returns a string reference
+    pub fn resolve<'a>(&self, pool: &'a PoolRef) -> Option<&'a str> {
+        // Safe because the new-type wraps the ffi::id and cant be created otherwise
+        unsafe {
+            let c_str = ffi::pool_id2str(pool.as_ptr().as_ptr(), self.0);
+            let c_str = CStr::from_ptr(c_str).to_str().expect("utf-8 parse error");
+            if c_str == "<NULL>" {
+                None
+            } else {
+                Some(c_str)
+            }
```

Comment:
```suggestion
            (c_str != "<NULL>").then_some(c_str)
```

Reply:
Such strange behavior, what if you would want to store the string `"<NULL>"`?

Should we maybe assert in the `intern` method that `"<NULL>"` is not a valid internable string?

Also should we turn `"<NULL>"` into a constant?